diff --git a/node_modules/@pulumi/pulumi/.bun-tag-46d0f34eb565d783 b/.bun-tag-46d0f34eb565d783
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/node_modules/@pulumi/pulumi/.bun-tag-48654133d060ad05 b/.bun-tag-48654133d060ad05
new file mode 100644
index 0000000000000000000000000000000000000000..e69de29bb2d1d6434b8b29ae775ad8c2e48c5391
diff --git a/cmd/run/index.js b/cmd/run/index.js
index 0757cf0d31363909f424a0aee2bcd95f264d1b9f..a5d53602878b12b4a44cf46b590babfdf4ae1da6 100644
--- a/cmd/run/index.js
+++ b/cmd/run/index.js
@@ -152,7 +152,6 @@ async function beforeExitHandler(code) {
 // fact.  For example, we want to keep track of ScriptId->FileNames so that we can appropriately
 // report errors for Functions we cannot serialize.  This can only be done (up to Node11 at least)
 // by register to hear about scripts being parsed.
-const v8Hooks = __importStar(require("../../runtime/closure/v8Hooks"));
 // This is the entrypoint for running a Node.js program with minimal scaffolding.
 const minimist_1 = __importDefault(require("minimist"));
 function usage() {
@@ -230,7 +229,8 @@ function main(args) {
     addToEnvIfDefined("PULUMI_NODEJS_ENGINE", argv["engine"]);
     addToEnvIfDefined("PULUMI_NODEJS_SYNC", argv["sync"]);
     // Ensure that our v8 hooks have been initialized.  Then actually load and run the user program.
-    v8Hooks.isInitializedAsync().then(() => {
+    const ready = process.versions.bun ? Promise.resolve() : import("../../runtime/closure/v8Hooks.js").then((v8Hooks) => v8Hooks.isInitializedAsync());
+    ready.then(() => {
         const promise = require("./run").run(argv, 
         /*programStarted:   */ () => {
             programRunning = true;
@@ -250,7 +250,7 @@ function main(args) {
 }
 function addToEnvIfDefined(key, value) {
     if (value) {
-        process.env[key] = value;
+        process.env[key] = `${value}`;
     }
 }
 main(process.argv.slice(2));
diff --git a/cmd/run/run.js b/cmd/run/run.js
index 2e2e1fd9dc9379197bc7a6c9f3e8b47e80c10a22..90a69c07f62c238c0ff36860700b45e8a5099457 100644
--- a/cmd/run/run.js
+++ b/cmd/run/run.js
@@ -235,7 +235,7 @@ async function run(argv, programStarted, reportLoggedError, isErrorReported) {
         process.chdir(pwd);
     }
     // If this is a typescript project, we'll want to load node-ts.
-    const typeScript = process.env["PULUMI_NODEJS_TYPESCRIPT"] === "true";
+    const typeScript = !process.versions.bun && process.env["PULUMI_NODEJS_TYPESCRIPT"] === "true";
     // We provide reasonable defaults for many ts options, meaning you don't need to have a tsconfig.json present
     // if you want to use TypeScript with Pulumi. However, ts-node's default behavior is to walk up from the cwd to
     // find a tsconfig.json. For us, it's reasonable to say that the "root" of the project is the cwd,
diff --git a/cmd/run-plugin/index.js b/cmd/run-plugin/index.js
index 5907d393cf8ac711624e3a77484adddba1c3158e..0c4dc3fe4f7895215619cfce77ddcf4068ed69e8 100644
--- a/cmd/run-plugin/index.js
+++ b/cmd/run-plugin/index.js
@@ -82,7 +82,6 @@ process.on("exit", (code) => {
 // fact.  For example, we want to keep track of ScriptId->FileNames so that we can appropriately
 // report errors for Functions we cannot serialize.  This can only be done (up to Node11 at least)
 // by register to hear about scripts being parsed.
-const v8Hooks = __importStar(require("../../runtime/closure/v8Hooks"));
 // This is the entrypoint for running a Node.js program with minimal scaffolding.
 const minimist_1 = __importDefault(require("minimist"));
 function usage() {
@@ -100,7 +99,8 @@ function main(args) {
         return printUsageAndExit();
     }
     // Ensure that our v8 hooks have been initialized.  Then actually load and run the user program.
-    v8Hooks.isInitializedAsync().then(() => {
+    const ready = process.versions.bun ? Promise.resolve() : import("../../runtime/closure/v8Hooks.js").then((v8Hooks) => v8Hooks.isInitializedAsync());
+    ready.then(() => {
         const promise = require("./run").run({
             argv,
             programStarted: () => {
diff --git a/cmd/run-policy-pack/index.js b/cmd/run-policy-pack/index.js
index 1c701d52af34e0a52f2dabca7c52e4137549ae8a..53b3a9d97b1c5cd0b9212484354c55a073bdcd5f 100644
--- a/cmd/run-policy-pack/index.js
+++ b/cmd/run-policy-pack/index.js
@@ -82,7 +82,6 @@ process.on("exit", (code) => {
 // fact.  For example, we want to keep track of ScriptId->FileNames so that we can appropriately
 // report errors for Functions we cannot serialize.  This can only be done (up to Node11 at least)
 // by register to hear about scripts being parsed.
-const v8Hooks = __importStar(require("../../runtime/closure/v8Hooks"));
 // This is the entrypoint for running a Node.js program with minimal scaffolding.
 const minimist_1 = __importDefault(require("minimist"));
 function usage() {
@@ -102,7 +101,8 @@ function main(args) {
     // Remove <engine-address> so we simply execute the program.
     argv._.shift();
     // Ensure that our v8 hooks have been initialized.  Then actually load and run the user program.
-    v8Hooks.isInitializedAsync().then(() => {
+    const ready = process.versions.bun ? Promise.resolve() : import("../../runtime/closure/v8Hooks.js").then((v8Hooks) => v8Hooks.isInitializedAsync());
+    ready.then(() => {
         const promise = require("./run").run({
             argv,
             programStarted: () => {
diff --git a/runtime/index.js b/runtime/index.js
index 3c616eed05fd57db76aa37e76c11507ec255ab75..6dd47a8a2b3e62e7a6f018c9f1f0d529ee3901a9 100644
--- a/runtime/index.js
+++ b/runtime/index.js
@@ -16,9 +16,14 @@ function __export(m) {
     for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
 }
 Object.defineProperty(exports, "__esModule", { value: true });
-var serializeClosure_1 = require("./closure/serializeClosure");
-exports.serializeFunctionAsync = serializeClosure_1.serializeFunctionAsync;
-exports.serializeFunction = serializeClosure_1.serializeFunction;
+exports.serializeFunctionAsync = async (...args) => {
+  var serializeClosure_1 = await import("./closure/serializeClosure");
+  return serializeClosure_1.serializeFunctionAsync(...args);
+};
+exports.serializeFunction = async (...args) => {
+  var serializeClosure_1 = await import("./closure/serializeClosure");
+  return serializeClosure_1.serializeFunction(...args);
+};
 var codePaths_1 = require("./closure/codePaths");
 exports.computeCodePaths = codePaths_1.computeCodePaths;
 var mocks_1 = require("./mocks");
